In this notebook, I will demonstrate how to solve **Time Value of Money** problems using the NumPy library.

In [1]:
import numpy as np
import platform
print('Python Version: {}'.format(platform.python_version()))
print('Numpy Version: {}'.format(np.__version__))

Python Version: 3.6.1
Numpy Version: 1.12.1


# Example 1: Simple Compounding
**A client invests $2 million in a money market fund projected to earn 3 percent annually. Estimate the value of this investment after seven years.**<br><br>
This is a very simple problem where we have to calculate the future value of an investment that grows over a certain period of time by virtue of compounding.

In [2]:
rate = 3 / 100      # Given that fund is projected to earn 3% annually
number_periods = 7  # Given that the client stays invested for 7 years
pmt = 0             # Zero because there are no intermediate payments
initial_investment = -2000000  # Negative because it's a cash outflow

fv_of_portfolio = np.fv(rate=rate, nper=number_periods, pmt=pmt, pv=initial_investment)

print('${:,.2f}'.format(fv_of_portfolio))

$2,459,747.73


As can be seen from the output above, the client’s initial investment of **\$2 million** will have risen to **\$2,459,747.73** after 7 years.
<br>
# Example 2: Calculating the Present Value of a future investment
**To cover the first year’s total college tuition payments for his two children, a father will make a \$75,000 payment five years from now. How much will he need to invest today to meet his first tuition goal if the investment earns 6 percent annually?**<br><br>
The amount that the father invests today grows to equal the required cash outflow of \$75,000 five years from now. Therefore, we have to calculate the present value of this cash outflow at the given 6% rate of return.

In [3]:
rate = 6 / 100      # Given that the investment earns 6% annually
number_periods = 5  # 5 years of investment
pmt = 0             # No intermediate payments; hence, zero
expected_value = 75000  # Expected value at end of investment period

present_value = np.pv(rate=rate, nper=number_periods, pmt=0, fv=expected_value)

print('${:,.2f}'.format(-present_value))

$56,044.36


Thus, the father needs to make an investment of **\$56,044.36** today to meet his goal.
<br>

# Example 3: Calculating the compound interest on an investment
**An investor contributes \$100,000 to an investment portfolio today. At the end of 5 years, she finds that the portfolio is now worth \$155,000. What has been the growth rate of her portfolio?**
<br><br>
This is easily obtained by calling the rate() function with the appropriate parameters.

In [4]:
number_periods = 5
pmt = 0
investment = -100000
fv = 155000

rate_of_growth = np.rate(nper=number_periods, pmt=pmt, pv=investment, fv=fv)

print('{:.2f}%'.format(rate_of_growth * 100))

9.16%


Therefore, the investor’s portfolio has grown at a compounded rate of **9.16%** per year for the last 5 years.<br>

# Example 4: Calculating monthly payments
**You are planning to purchase a \$600,000 house by making a down payment of \$120,000 and borrowing the remainder with a 30-year fixed-rate mortgage with monthly payments. The first payment is due at the end of the first year, i.e., t = 1. Current mortgage interest rates are quoted at 2.69 percent with monthly compounding. What will your monthly mortgage payments be?**

In [5]:
rate = (2.59 / 12) / 100  # Divide by 12 to account for monthly compounding
nper = 30 * 12  # Monthly payments each year for 30 years

# Loan amount = Purchase price less down payment; negative sign to indicate cash outflow
loan_amount = -(600000 - 120000)  

future_value = 0  # Zero because the loan is fully paid off at the end of the 30th year

monthly_payment = np.pmt(rate=rate, nper=nper, pv=loan_amount, fv=future_value)

print('${:,.2f}'.format(monthly_payment))

$1,919.12


As can be seen above, the monthly payment for this mortgage structure will be **\$1,919.12**.<br>

# Example 5: The Projected Annuity Amount Needed to Fund a Future-Annuity Inflow

**Jill Grant is 22 years old (at t = 0) and is planning for her retirement at age 63 (at t = 41). She plans to save \$2,000 per year for the next 15 years (t = 1 to t = 15). She wants to have retirement income of \$100,000 per year for 20 years, with the first retirement payment starting at t = 41. How much must Grant save each year from t = 16 to t = 40 in order to achieve her retirement goal? Assume she plans to invest in a diversified stock-and-bond mutual fund that will earn 8 percent per year on average.**
<br><br>

Here, we have to find out how much Grant needs to contribute to her retirement portfolio every year for the 25 years from t=16 to t=40 so that she can meet her retirement goal.

We will solve this problem by breaking it down into the following three steps:
1.  Find the value of Grant’s portfolio at the end of t=15. This is a simple future value calculation.
+ The next step is to calculate the final value of the portfolio right before Grant start funding her retirement. This is the same as the present value of the annuity of \$100,000 per year for 20 years at t=40.
+ Now that we have the portfolio values at t=15 and t=40, we have to calculate the annual payments for 25 years. For this calculation, the portfolio values at t=15 and t=40 become the present value and future value of the portfolio, respectively.

In [6]:
# With $2000 saved per year for the next 15 years,
# let's see what the portfolio would be worth at the end of these 15 years
portfolio_t_15 = np.fv(rate=8/100, nper=15, pmt=-2000, pv=0)
print('${:,.2f}'.format(portfolio_t_15))

$54,304.23


Therefore, the portfolio value at t=15 will be **\$54,304.23**.

In [7]:
# Now, let's calculate the required value of her portfolio at retirement
# to fund an annuity of $100,000 per year for 20 years
# This can be done by computing the present value of the annuity
portfolio_t_40 = np.pv(rate=8/100, nper=20, pmt=-100000, fv=0)
print('${:,.2f}'.format(portfolio_t_40))

$981,814.74


Thus, to fund the annuity, there should be **\$981,814.74** in the retirement account.

In [8]:
# Now, the last step is to determine how much Grant needs to contribute
# to her portfolio annually from t=15 to t=40, i.e, for a total of 25 years
reqd_savings = np.pmt(rate=8/100, nper=25, pv=portfolio_t_15, fv=-portfolio_t_40)
print('${:,.2f}'.format(reqd_savings))

$8,342.87


Thus, we see that Grant needs to contribute **\$8,342.87** every year from t=16 to t=40 to meet her retirement goals.